docs/INGESTION.md

```diff
@@ -0,0 +1,99 @@
+```rust
+pub struct Datum<'a> {
+    timeseries_id: i64,
+    // needed for QC
+    param_id: i32,
+    value: ObsType<'a>,
+    qc_usable: bool,
+}
+
+/// Generic container for a piece of data ready to be inserted into the DB
+pub struct DataChunk<'a> {
+    timestamp: DateTime<Utc>,
+    time_resolution: Option<chronoutil::RelativeDuration>,
+    data: Vec<Datum<'a>>,
+}
+```
+
+4. Inserting: The data then needs to be QCed and inserted into the database. As it is already in a generic format, we provide a function that can be used for this purpose `qc_and_insert_data`
+
+An example of this coming together in an HTTP handler:
+```rust
+async fn handle_kldata(
+    State(pools): State<DbPools>,
+    State(param_conversions): State<ParamConversions>,
+    State(permit_table): State<Arc<RwLock<(ParamPermitTable, StationPermitTable)>>>,
+    State(rove_connector): State<Arc<rove_connector::Connector>>,
+    State(qc_pipelines): State<Arc<HashMap<(i32, RelativeDuration), rove::Pipeline>>>,
+    body: String,
+) -> Json<KldataResp> {
+    metrics::counter!(KLDATA_MESSAGES_RECEIVED).increment(1);
+
+    let result: Result<usize, Error> = async {
+        let mut open_conn = pools.open.get().await?;
+        let mut restricted_conn = pools.restricted.get().await?;
+
+        let (message_id, obsinn_chunk) = parse_kldata(&body, param_conversions.clone())?;
+
+        let (mut open_data, mut restricted_data) = filter_and_label_kldata(
+            obsinn_chunk,
+            &mut open_conn,
+            &mut restricted_conn,
+            param_conversions,
+            permit_table,
+        )
+        .await?;
+
+        qc_and_insert_data(
+            &mut open_data,
+            &rove_connector,
+            &qc_pipelines,
+            &mut open_conn,
+        )
+        .await?;
+        qc_and_insert_data(
+            &mut restricted_data,
+            &rove_connector,
+            &qc_pipelines,
+            &mut restricted_conn,
+        )
+        .await?;
+
+        Ok(message_id)
+    }
+    .await;
+
+    match result {
+        Ok(message_id) => Json(KldataResp {
+            message: "".into(),
+            message_id,
+            res: 0,
+            retry: false,
+        }),
+        Err(e) => {
+            metrics::counter!(KLDATA_FAILURES).increment(1);
+            error!("failed to ingest kldata message: {}, body: {}", e, body);
+            // TODO: log errors?
+            Json(KldataResp {
+                message: e.to_string(),
+                message_id: 0, // TODO: some clever way to get the message id still if possible?
+                res: 1,
+                retry: !matches!(e, Error::Parse(_)),
+            })
+        }
+    }
+}
+```
```

Comment:
Should this just be a link to the lines in the file? We have to fix it here too when the underlying code changes? But I also kinda like seeing the code directly in the README :thinking:

Reply:
I think linking the lines will be too fragile, IMO it's a better failure mode to have the snippet be outdated than incomprehensible